README.md
# fabrik

![Build Status](https://github.com/skovmand/fabrik/workflows/Fabrik%20CI/badge.svg)

A Sudoku solver with backtracking written in Rust ❤️

[![asciicast](https://asciinema.org/a/434516.svg)](https://asciinema.org/a/434516)


## Setup

1. [Install Rust](https://www.rust-lang.org/tools/install)
2. Clone this repository using `git@github.com:skovmand/fabrik.git`
3. Build the binary using `cargo build --release`
4. Copy `./target/release/fabrik` to wherever you like


## Usage

Put a sudoku in a file, with empty fields as `-` and `1-9` as field values. For example

```
-3-----8-
5-------4
--42-81--
1-34-92-5
---------
4-68-53-9
--17-35--
9-------1
-6-----7-
```

If the file is called `sudoku-1.txt`, solve it using `fabrik <filename>`. If you pass the `--display` or `-d` flag, the sudoku will be solved with a delay after each step. This is great for learning how backtracking works.

```
# Solve a sudoku in display mode
$ fabrik sudoku-1.txt --display

# Solve a sudoku in display with custom delay (1ms per step)
$ fabrik sudoku-1.txt --display --delay 1

# Solve a sudoku and display the final result
$ fabrik sudoku-1.txt
+-----------+
|632|514|987|
|518|397|624|
|794|268|153|
+---+---+---+
|183|479|265|
|259|136|748|
|476|825|319|
+---+---+---+
|821|743|596|
|947|652|831|
|365|981|472|
+-----------+
```


## Tests

Run the project tests using `cargo test`


## Benchmarks

Run the project benchmarks using `cargo bench`

Results on my MacBook Air M1 2020 (256GB/8GB ram). The "hard sudoku" is a newspaper sudoku with level "hard". The "very hard sudoku" is the hardest one my father-in-law knows :-)

```
solve hard sudoku (1432-3)
                        time:   [146.26 us 146.30 us 146.36 us]

solve very hard sudoku (from ykw1)
                        time:   [365.79 us 365.89 us 366.01 us]
```


CHANGELOG.md
## 0.6.3 - Refactoring and slight speedup

* Implement IntoIterator for SudokuBoard, #15 #16
* Functional implementations of row, column, square searches
* Remember current position instead of iterating the whole board for empty fields #11
* Use Position for board positions #10

## 0.6.2 - Small optimizations

* Faster row lookups #8
* Only create SudokuField::Value once #9

## 0.6.1 - Delay flag + internal refactors

* Use --delay flag for custom render speed in display mode
* Internal refactors: Use generics instead of dynamic dispatch

## 0.6.0 - Visualizing solutions

* Visualize sudoku solving when using the --display or -d flag

## 0.5.2 - Benchmarks and polishing

* Add benchmark suite using criterion
* Add README.md

## 0.5.1 - Custom error handling

* Remove anyhow, handle errors manually (learn you some error handling for great good)
* Remove an unused old test block

## 0.5.0 - First Release


src/lib.rs
pub mod renderers;
pub mod sudoku;

use renderers::SudokuRenderer;
use sudoku::{Position, SudokuBoard, SudokuError, SudokuField};

/// Solve a SudokuBoard with callback to a SudokuRenderer at every step
pub fn solve_board<T: SudokuRenderer>(
    board: &mut SudokuBoard,
    renderer: &T,
) -> Result<SudokuBoard, SudokuError> {
    do_solve_board(board, renderer, Position { row: 0, column: 0 })
}

// Solve the Sudoku Board
// Backtracking here is at once very advanced and also really simple!
// For every iteration we are finding the first empty field on the board,
// scanning row, column and square to check if a number is valid,
// then placing the number in the field. After this we will recurse by calling
// the same function once again, but below the recursion we will put back None
// into the field, so if the program backtracks, it will restore the original state.
fn do_solve_board<T: SudokuRenderer>(
    board: &mut SudokuBoard,
    renderer: &T,
    position: Position,
) -> Result<SudokuBoard, SudokuError> {
    for num in 1..=9 {
        let field = SudokuField::Value(num);

        if board.valid_number(&position, &field) {
            board.put_field(&position, field);
            renderer.display_step(board);

            if let Some(position) = board.next_empty_field(&position) {
                // Another free field is present, let's jump to that and continue solving
                if let Ok(board) = do_solve_board(board, renderer, position) {
                    return Ok(board);
                }

                // We met a dead end, set the field to Empty again (after which we will continue looping)
                board.put_field(&position, SudokuField::Empty);
            } else {
                // The sudoku has no more empty fields, we are done!
                return Ok(board.clone());
            };
        }
    }

    Err(SudokuError::Unsolvable)
}

#[cfg(test)]
mod tests {
    use std::convert::TryFrom;

    use crate::renderers::NonRenderer;

    use super::*;
    use indoc::indoc;

    #[test]
    fn solves_board() {
        const TEST_SUDOKU: &str = indoc! {"
            -47---96-
            8--716--2
            6-------8
            --21-85--
            ----9----
            --16-23--
            5-------1
            7--945--3
            -69---75-
        "};

        let mut board = SudokuBoard::try_from(TEST_SUDOKU.to_owned()).unwrap();
        solve_board(&mut board, &NonRenderer {}).expect("Could not solve test board");

        let expected_solution = indoc! {"
            +-----------+
            |147|823|965|
            |895|716|432|
            |623|459|178|
            +---+---+---+
            |972|138|546|
            |386|594|217|
            |451|672|389|
            +---+---+---+
            |534|267|891|
            |718|945|623|
            |269|381|754|
            +-----------+
        "};

        assert_eq!(board.to_string(), expected_solution);
    }

    #[test]
    fn fails_on_invalid_board() {
        const INVALID_SUDOKU: &str = indoc! {"
            -47---96-
            84-716--2
            6-------8
            --21-85--
            ----9----
            --16-23--
            5-------1
            7--945--3
            -69---75-
        "};

        let mut board = SudokuBoard::try_from(INVALID_SUDOKU.to_owned()).unwrap();
        let result = solve_board(&mut board, &NonRenderer {});

        assert!(result.is_err());
        assert_eq!(result.err().unwrap(), SudokuError::Unsolvable);
    }

    #[test]
    fn solves_blank_board() {
        const BLANK_SUDOKU: &str = indoc! {"
            ---------
            ---------
            ---------
            ---------
            ---------
            ---------
            ---------
            ---------
            ---------
        "};

        let mut board = SudokuBoard::try_from(BLANK_SUDOKU.to_owned()).unwrap();
        solve_board(&mut board, &NonRenderer {}).expect("Could not solve test board");

        let expected_solution = indoc! {"
            +-----------+
            |123|456|789|
            |456|789|123|
            |789|123|456|
            +---+---+---+
            |214|365|897|
            |365|897|214|
            |897|214|365|
            +---+---+---+
            |531|642|978|
            |642|978|531|
            |978|531|642|
            +-----------+
        "};

        assert_eq!(board.to_string(), expected_solution);
    }

    #[test]
    fn solves_hard_board() {
        const HARD_SUDOKU: &str = indoc! {"
                -3-----8-
                5-------4
                --42-81--
                1-34-92-5
                ---------
                4-68-53-9
                --17-35--
                9-------1
                -6-----7-
        "};

        let mut board = SudokuBoard::try_from(HARD_SUDOKU.to_owned()).unwrap();
        solve_board(&mut board, &NonRenderer {}).expect("Could not solve test board");

        let expected_solution = indoc! {"
            +-----------+
            |632|514|987|
            |518|397|624|
            |794|268|153|
            +---+---+---+
            |183|479|265|
            |259|136|748|
            |476|825|319|
            +---+---+---+
            |821|743|596|
            |947|652|831|
            |365|981|472|
            +-----------+
        "};

        assert_eq!(board.to_string(), expected_solution);
    }
}
